src/CP/valueselection/learning/supervisedlearnedheuristic.jl

```diff
@@ -0,0 +1,143 @@
+    search_metrics::Union{Nothing,SearchMetrics}
+    firstActionTaken::Bool
+    trainMode::Bool
+    helpVariableHeuristic::AbstractVariableSelection
+    helpValueHeuristic::ValueSelection
+    helpSolution::Union{Nothing,Solution}
+    eta_init::Float64
+    eta_stable::Float64
+    warmup_steps::Int64
+    decay_steps::Int64
+    step::Int64
+    rng::AbstractRNG #for reproducibility
+    
+    function SupervisedLearnedHeuristic{SR,R,A}(
+        agent::RL.Agent;
+        helpVariableHeuristic::AbstractVariableSelection=MinDomainVariableSelection(),
+        helpValueHeuristic::ValueSelection=BasicHeuristic(),
+        eta_init::Float64=0.5,
+        eta_stable::Float64=0.5,
+        warmup_steps::Int64=0,
+        decay_steps::Int64=0,
+        rng::AbstractRNG=MersenneTwister()
+    ) where {SR,R,A}
+        new{SR,R,A}(agent, nothing, nothing, nothing, nothing, nothing, nothing, false, true, helpVariableHeuristic, helpValueHeuristic, nothing, eta_init, eta_stable, warmup_steps, decay_steps, 0, rng)
+    end
+end
+
+"""
+    (valueSelection::SupervisedLearnedHeuristic)(::InitializingPhase, model::CPModel, x::Union{Nothing, AbstractIntVar}, current_status::Union{Nothing, Symbol})
+
+Update the part of the SupervisedLearnedHeuristic which act like an RL environment, by initializing it with informations caracteritic of the new CPModel. Creates an
+initial observation with a false variable. A fixPoint! will be called before the SupervisedLearnedHeuristic takes its first decision.
+Finally, makes the agent call the process of the RL pre_episode_stage (basically making sure that the buffer is empty).
+"""
+function (valueSelection::SupervisedLearnedHeuristic)(::Type{InitializingPhase}, model::CPModel)
+    # create the environment
+    valueSelection.firstActionTaken = false
+    update_with_cpmodel!(valueSelection, model)
+    # FIXME get rid of this => prone to bugs
+    false_x = first(values(branchable_variables(model)))
+    env = get_observation!(valueSelection, model, false_x)
+    
+    eta = get_eta(valueSelection) #get the current eta_init 
+    if rand(valueSelection.rng) < eta
+        #the instance is solved using classic CP on a duplicated model
+        model_duplicate = deepcopy(model) 
+        strategy = DFSearch()
+        search!(model_duplicate, strategy, valueSelection.helpVariableHeuristic, valueSelection.helpValueHeuristic)
+        #the solution is added to the valueSelection.helpSolution field.
+        if !isnothing(model_duplicate.statistics.solutions)
+            solutions = model_duplicate.statistics.solutions[model_duplicate.statistics.solutions.!=nothing]
+            if length(solutions) >= 1
+                valueSelection.helpSolution = solutions[1]
+            end
+        end
+        reset_model!(model_duplicate)
+    else
+        valueSelection.helpSolution = nothing
+    end
+
+    # Reset the agent, useful for things like recurrent networks
+    Flux.reset!(valueSelection.agent)
```

Comment:
I'm just wondering why we previously added this line?

Reply:
I honestly have no idea, it seemed to me that this was not a problem but I don't know if it is useful.